### DISCUSSION TITLE
Create simplified version of HTTP Endpoint

### DISCUSSION THREAD

**Discussion by tomy2105** [elsa-workflows/elsa-core#1249] [Q&A] [answered]
Elsa 2.0 from NuGet....
Inspired by your answer to https://github.com/elsa-workflows/elsa-core/discussions/1212 and based on the `WebhookActivityTypeProvider` I tried adding my simple HTTP Endpoint based activity using code below.

Activity is visible in designer, however calling its URL/path does not trigger workflow.
Similar as #1237 I don't see `ActivateActivityAsync` ever invoked.

Can you advise what I am missing?
Do I need my bookmark provider (your answer in https://github.com/elsa-workflows/elsa-core/discussions/1212 seems to indicate it is not required but I might have misunderstood it :()?

Thanks in advance.

    public class EndpointActivityProvider : IActivityTypeProvider
    {
        private readonly IActivityActivator _activityActivator;

        public EndpointActivityProvider(
            IActivityActivator activityActivator)
        {
            _activityActivator = activityActivator;
        }

        public async ValueTask<IEnumerable<ActivityType>> GetActivityTypesAsync(CancellationToken cancellationToken = default)
        {
            var activityTypes = new List<ActivityType>();
            var activity = CreateEndpointActivityType();
            activityTypes.Add(activity);

            return activityTypes;
        }

        private ActivityType CreateEndpointActivityType()
        {
            var descriptor = new ActivityDescriptor
            {
                Type = "TestTypeName",
                DisplayName = "TestDisplayName",
                Category = "HTTP",
                Outcomes = new[] { OutcomeNames.Done },
                Traits = ActivityTraits.Trigger,
                Properties = new[]
                {
                    new ActivityPropertyDescriptor(
                        nameof(HttpEndpoint.Path),
                        typeof(string),
                        ActivityPropertyUIHints.SingleLine,
                        "Path",
                        "Specify Path",
                        null,
                        null,
                        "",
                        SyntaxNames.Literal,
                        new[] { SyntaxNames.Literal, SyntaxNames.JavaScript, SyntaxNames.Liquid })
                }
            };

            async ValueTask<IActivity> ActivateActivityAsync(ActivityExecutionContext context)
            {
                var activity = await _activityActivator.ActivateActivityAsync<HttpEndpoint>(context);

                activity.Methods = new HashSet<string>(new[] { "GET" });
                activity.ReadContent = false;
                activity.TargetType = null;
                return activity;
            }

            return new ActivityType
            {
                TypeName = descriptor.Type,
                Type = typeof(HttpEndpoint),
                Description = descriptor.Description,
                DisplayName = descriptor.DisplayName,
                ActivateAsync = ActivateActivityAsync,
                Describe = () => descriptor
            };
        }
    }

**Comment by sfmskywalker** [comment]
> Do I need my bookmark provider (your answer in #1212 seems to indicate it is not required but I might have misunderstood it :()?

Indeed, in this case you should provide a custom bookmark provider. The `HttpEndpointMiddleware` is wired to find all bookmarks that are associated with the activity type name `HttpEndpoint`.

However, you're providing an activity with type name `TestTypeName`. Writing a custom bookmark provider allows you to create bookmarks that will be found by the HTTP Endpoint Middleware component.

Here's an example:

```csharp

// The custom provider must still yield HttpEndpointBookmark bookmark objects - this is what the middleware expects.
public class EndpointBookmarkProvider : BookmarkProvider<HttpEndpointBookmark>
{
    // Tell the bookmark finder that we support "TestTypeName" activities.
    public override bool SupportsActivity(BookmarkProviderContext context) => context.ActivityType.TypeName == "TestTypeName";

    public override async ValueTask<IEnumerable<BookmarkResult>> GetBookmarksAsync(BookmarkProviderContext context, CancellationToken cancellationToken)
    {
        // We need to create an HttpEndpointBookmark with a Path to match against.
        var path = ToLower(await context.ReadActivityPropertyAsync<HttpEndpoint, PathString>(x => x.Path, cancellationToken))!;

        // As well as a list of supported HTTP methods. These need to normalized into lowercase representations.
        var methods = new [] { "get" };
        
        return new[] { Result(new(path, methods, nameof(HttpEndpoint)) };
    }

    private static string ToLower(string s) => s.ToLowerInvariant();
}
```

When you register the above with `AddBookmarkProvider<EndpointBookmarkProvider>()`, it should work.

**Comment by tomy2105** [comment]
Ok, will try, however another question..
If I need to do my own bookmark provider, what is the benefit of using this IActivityTypeProvider approach instead of inheriting (or similar) from HTTPEndpoint?

**Reply by sfmskywalker** [reply]
The only reasons I can think of to using a custom activity type provider are:

* You have a dynamic source of information from which you want to project HTTP Endpoint activity variations (like the Webhooks module does).
* You want to customize what properties are available and perhaps even the property details themselves.

But if you only need a fixed set of custom HTTP endpoint activity types, you might probably just as well derive from HTTP endpoint directly.

Now I have a question for you :)

What are possible reasons for someone to create their own version of the HTTP Endpoint activity, instead of simply using the HTTP Endpoint activity as-is?

**Reply by tomy2105** [reply]
Oh, the answer to that is simple.
My use case is limited to "POST" only requests which always need read content flag and target type is always set to same value (which since my class is in different assembly consists of full type name and full assembly name).

The person who will actually be writing the workflows is non-techincal and wants to spend as little time as possible doing this so it has to be made as simple as possible to him.
So I want HTTP Endpoint which hides everything (and hardcodes to appropriate values) except path, so the person in question doesn't have to type/copy/paste all other stuff (makes errors in the process and wake me up because something doesn't work :)).

**Reply by sfmskywalker** [reply]
That makes perfect sense 😄

**Comment by tomy2105** [comment]
In addition my compiler cannot find ReadActivityPropertyAsync? Is this by any chance added in 2.1 or......?

**Reply by sfmskywalker** [reply]
Ah, yes I believe that was added in 2.1. If you want to make this work in 2.0, I think you can just copy the implementation of that method (and remove it when you do update to 2.1).

**Reply by tomy2105** [reply]
And Result? Is that BookmarkResult which also seems to be added in 2.1?

**Reply by sfmskywalker** [reply]
Yes, `BookmarkResult` was added in 2.1 as well.

**Comment by tomy2105** [comment]
OK, so I switched to trying to do this with extending HTTPEndpoint and code is below.....
But it still doesn't work....
Still in 2.0 code, remember!
My `GetBookmarksAsync` is invoked and I can trace that trigger is create but with activity type set to `TestReceiveRequest`.
And since `HttpEndpointMiddleware` search requires activity type set to `HttpEndpoint` that is probably the reason why it doesn't work.

Is there a way to set activity type to `HttpEndpoint` in 2.0 version (your code above indicates usage of `BookmarkResult` which seems to be added in 2.1?

    [Trigger(
        Category = "Test",
        DisplayName = "Receive Request",
        Description = "Wait for an test request.",
        Outcomes = new[] { OutcomeNames.Done }
    )]
    public class TestReceiveRequest : HttpEndpoint
    {
        [ActivityProperty(Hint = "The relative path for the test request.", SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid })]
        public new PathString Path { get; set; }

        public new HashSet<string> Methods { get; set; } = new(new[] { "GET" });

        public new bool ReadContent { get; set; }

        public new Type? TargetType { get; set; }
    }

    public class TestReceiveRequestBookmarkProvider : BookmarkProvider<HttpEndpointBookmark, TestReceiveRequest>
    {
        public override async ValueTask<IEnumerable<IBookmark>> GetBookmarksAsync(BookmarkProviderContext<TestReceiveRequest> context, CancellationToken cancellationToken)
        {
            var path = ToLower(await context.Activity.GetPropertyValueAsync(x => x.Path, cancellationToken))!;
            var correlationId = ToLower(context.ActivityExecutionContext.WorkflowExecutionContext.CorrelationId);

            return new[] { new HttpEndpointBookmark(path, "GET", correlationId) };
        }

        private static string ToLower(string? s) => s?.ToLowerInvariant();
    }

**Reply by sfmskywalker** [reply]
That’s exactly the reason 2.1 let’s you provide a BookmarkResult that allows you to specify the activity type specifically :)

**Reply by tomy2105** [reply]
Switched, for test, to Elsa 2.1 and I got the bookmark and activity working using code below, good changes in 2.1! :).

**However** :(, the read content flag and target type don't work because they are not activity inputs properties (because I want them hidden from end user) but `HTTPEndpointMiddleware` mandates/wants them to be input properties because it fetches their value using `EvaluatePropertyValueAsync` :(.

**Is there a way to make this work without them being visible in the designer?**
**A flag for property not being displayed in UI, or something? Anything? :)**


    [Trigger(
        Category = "Test",
        DisplayName = "Receive Request",
        Description = "Wait for an test request.",
        Outcomes = new[] { OutcomeNames.Done }
    )]
    public class TestReceiveRequest : HttpEndpoint
    {
        [ActivityInput(Hint = "The relative path for the test request.", SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid })]
        public new PathString Path { get; set; }

        public new HashSet<string> Methods { get; set; }

        public new bool ReadContent { get; set; } = true;

        public new  Type? TargetType { get; set; }
    }

    public class TestReceiveRequestBookmarkkProvider : BookmarkProvider<HttpEndpointBookmark, TestReceiveRequest>
    {
        public override async ValueTask<IEnumerable<BookmarkResult>> GetBookmarksAsync(BookmarkProviderContext<TestReceiveRequest> context, CancellationToken cancellationToken)
        {
            var path = ToLower(await context.ReadActivityPropertyAsync<HttpEndpoint, PathString>(x => x.Path, cancellationToken))!;
            var methods = new[] { "post" };

            BookmarkResult CreateBookmark(string method) => Result(new(path, method), nameof(HttpEndpoint));
            return methods.Select(CreateBookmark);
        }

        private static string ToLower(string? s) => s?.ToLowerInvariant();
    }

**Reply by sfmskywalker** [reply]
> Switched, for test, to Elsa 2.1 and I got the bookmark and activity working using code below, good changes in 2.1! :).

Thanks 😄 

> However :(, the read content flag and target type don't work because they are not activity inputs properties (because I want them hidden from end user) but HTTPEndpointMiddleware mandates/wants them to be input properties because it fetches their value using EvaluatePropertyValueAsync :(.

What specifically won't work in this case?

> Is there a way to make this work without them being visible in the designer?

Right now, there is a way to **remove** the properties from the activity descriptor which might work for your case?

It requires that you implement a notification handler for the `DescribingActivityType` event, something like this:

```csharp
public class CustomizeTestReceiveRequest : INotificationHandler<DescribingActivityType>
{
    public Task Handle(DescribingActivityType notification, CancellationToken cancellationToken)
    {
        var descriptor = notification.ActivityDescriptor;
        descriptor.InputProperties = descriptor.InputProperties.Where(x => x.Name != "ReadContent " && x.Name != "TargetType ").ToArray();
        return Task.CompletedTask;
    }
}
```

Should work. But indeed, a nicer options would be to have a visibility flag to instruct the designer whether the property is hidden, readonly or visible (which would be the default).

**Reply by tomy2105** [reply]
> What specifically won't work in this case?

It behaves as if `ReadContent` is `false` and `TargetType` is `emtpy/null`, I always get `HTTPRequestModel` with Body set to `null`.

Will try notification handler and let you know if it works for my use case.....

**Reply by tomy2105** [reply]
Tried notification handler and indeed it allows me to hide properties from UI while having them in code.
But the code (see below) where I set new default values for ReadContent and TargetType still doesn't work (it behaves as if `ReadContent` is false and `TargetType` is emtpy/null, I always get `HTTPRequestModel` with Body set to `null`).

It seems that `HTTPEndpointMiddleware` is ignoring my default values in code :(.
Probably this two lines:

            var activityWrapper = workflowBlueprintWrapper.GetUnfilteredActivity<HttpEndpoint>(pendingWorkflow.ActivityId!);
            var readContent = await activityWrapper!.EvaluatePropertyValueAsync(x => x.ReadContent, cancellationToken);

I've placed breakpoint on `ReadContent` setter and getter both in my code and in original Elsa code and they get invoked way after `HTTPEndpointMiddleware` is executing (when the workflow is running). 
I've even tried making `ReadContent` virtual in Elsa code and override in mine, but that did not change anything.

It seems to me `HTTPEndpointMiddleware` is ignoring default values and code in this case :(.....
What can I do to make it work? :(.


    public class CustomizeTestReceiveRequest : INotificationHandler<DescribingActivityType>
    {
        // hiding some properties from UI
        public Task Handle(DescribingActivityType notification, CancellationToken cancellationToken)
        {
            var descriptor = notification.ActivityDescriptor;
            if (descriptor.Type == typeof(TestReceiveRequest).Name)
            {
                descriptor.InputProperties = descriptor.InputProperties.Where(x => x.Name != "ReadContent" && x.Name != "TargetType" && x.Name != "Methods").ToArray();
            }
            return Task.CompletedTask;
        }
    }

    [Trigger(
        Category = "Test",
        DisplayName = "Receive Request",
        Description = "Wait for an test request.",
        Outcomes = new[] { OutcomeNames.Done }
    )]
    public class TestReceiveRequest : HttpEndpoint
    {
        // here to specify default value
        [ActivityInput(DefaultValue = true)]
        public new bool ReadContent { 
            get; 
            set; 
        } = true;

        // here to specify default value
        [ActivityInput(DefaultValue = typeof(string))]
        public new Type? TargetType { get; set; } = typeof(string);
    }

    public class TestReceiveRequestBookmarkkProvider : BookmarkProvider<HttpEndpointBookmark, TestReceiveRequest>
    {
        public override async ValueTask<IEnumerable<BookmarkResult>> GetBookmarksAsync(BookmarkProviderContext<TestReceiveRequest> context, CancellationToken cancellationToken)
        {
            var path = ToLower(await context.ReadActivityPropertyAsync<HttpEndpoint, PathString>(x => x.Path, cancellationToken))!;
            var methods = new[] { "post" };

            BookmarkResult CreateBookmark(string method) => Result(new(path, method), nameof(HttpEndpoint));
            return methods.Select(CreateBookmark);
        }

        private static string ToLower(string? s) => s?.ToLowerInvariant();
    }

**Reply by tomy2105** [reply]
I've tried going through debugger from `HTTPEndpointMiddleware` and the problem seems in the fact that `activityWrapper!.EvaluatePropertyValueAsync` doesn't take into account default values, it only "sees" values that were set through designer.

I think `WorkflowBlueprintExtensions` `GetActivityPropertyValue` should take into account default values, however cannot get that code to compile because at that point I don't have IActivity derived class, not its properties so cannot access attrbviutes etc...
This is what I've tried (`var value = defaultValueResolver.GetDefaultValue(property);` doesn't compile :().....

        public static async ValueTask<object?> GetActivityPropertyValue(
            this IWorkflowBlueprint workflowBlueprint,
            string activityId,
            string propertyName,
            ActivityExecutionContext activityExecutionContext,
            CancellationToken cancellationToken)
        {
            var provider = workflowBlueprint.ActivityPropertyProviders.GetProvider(activityId, propertyName);

            if (provider == null)
            {
                var defaultValueResolver = activityExecutionContext.GetService<IActivityPropertyDefaultValueResolver>();
                var value = defaultValueResolver.GetDefaultValue(property);
                if (value != null)
                    return value;

                return default!;
            }

            var value = await provider.GetValueAsync(activityExecutionContext, cancellationToken);
            return value!;
        }

**Reply by sfmskywalker** [reply]
@tomy2105 Sorry for late response. Do you still need help with this?

**Reply by tomy2105** [reply]
@sfmskywalker  thanks I believe we covered everything in #1269.
My fork at https://github.com/tomy2105/elsa-core/blob/test/HTTPEndpointExtension/src/samples/dashboard/aspnetcore/ElsaDashboard.Samples.AspNetCore.Monolith/TestReceiveRequest.cs now works with all the changes you did for 2.2 release.

**Comment by tomy2105** [comment, marked as answer]
Code sample implementing custom simplified version of HTTP Endpoint can be seen at https://github.com/tomy2105/elsa-core/blob/test/HTTPEndpointExtension/src/samples/dashboard/aspnetcore/ElsaDashboard.Samples.AspNetCore.Monolith/TestReceiveRequest.cs.

Maybe someone wants to make a sample out of it :)?
